### ISSUE TITLE
Do not use TerminateThread to stop threads

### ISSUE THREAD

**Issue by lassoan** [PlusToolkit/PlusLib#719]
This change can introduce hard-to-detect regressions in PlusServer: https://github.com/PlusToolkit/PlusLib/pull/718

TerminateThread is a nuclear bomb - it is essentially a forced partial crash of the application. We should only use it as a very last resort, if ever, as it can put the application into an inconsistent state. It might be OK to do forced termination if there does not seem to be a chance for the thread to recover (for example, if a thread does not respond for 10 seconds; it may be still better to kill that thread instead of making the whole application hang indefinitely).

However, in this case, most likely the thread exits as requested, but for some reason vtkMultiThreader still sees the stopped thread as active.

**Comment by lassoan** [comment]
@adamaji Could you debug into this and see vtkMultiThreader did not see the thread as stopped before? Did the thread did not exit successfully? Or maybe we should just call CloseHandle(threadID) so that vtkMultiThreader can properly detect that the thread is completed?

**Comment by lassoan** [comment]
I've just checked and vtkMultiThreader::TerminateThread does not seem to actually terminate the thread but sets a flag that requests it to stop and waits for it to complete. So, the thread is actually not terminated forcefully (no risk of making the application to get into inconsistent state).

@adamaji can you check if keeping the original implementation (without https://github.com/PlusToolkit/PlusLib/pull/718) and just calling CloseHandle for the complete thread fixes the issue that you had?

**Comment by adamaji** [comment]
>can you check if keeping the original implementation (without #718) and just calling CloseHandle for the complete thread fixes the issue that you had?

Yes, I'll look into this

**Comment by adamaji** [comment]
From earlier debugging, the problem was that vtkMultiThreader did not see the thread as stopped, and thus does not free up its slots for `SpawnedThreadActiveFlag/Lock`. The DataReceiverThread in OpenIGTLinkServer does actually finish, setting `ClientData->DataReceiverActive.second` to false.

Using `CloseHandle` here doesn't work because we don't have any reference to the actual process id of the thread, the `DataReceiverThreadId` is just an index in the vtkMultiThreader. That's why I used vtkMultiThreader::TerminateThread in the first place -- it doesn't seem like there are other methods of freeing up the spawned thread variables in the multi threader without deconstructing entirely. We also cannot get the process id from the multi threader as those are out of scope / protected.

**Comment by lassoan** [comment]
Thanks for checking this. The followings should be sufficient then:

1. Revert back
 https://github.com/PlusToolkit/PlusLib/blob/f1fab0220b4d9420155f55068bcad70129669f7c/src/PlusServer/vtkPlusOpenIGTLinkServer.cxx#L1131

to 

    if (clientIterator->DataReceiverThreadId > 0)

because this means that as far as Plus concerned, the thread is still active (we don't rely on vtkMultiThreader internals).

2. Change

https://github.com/PlusToolkit/PlusLib/blob/f1fab0220b4d9420155f55068bcad70129669f7c/src/PlusServer/vtkPlusOpenIGTLinkServer.cxx#L1140-L1141

to 

```
// thread stopped
// We still need to call vtkMultiThreader::TerminateThread, not to terminate the thread, as it is already stopped,
// but to indicate that the thread ID can be reused. Without this, vtkMultiThreader would run out of usable thread IDs
// after a number of connects/disconnects.
this->Threader->TerminateThread(clientIterator->DataReceiverThreadId);
clientIterator->DataReceiverThreadId = -1;
```

**Comment by adamaji** [comment]
Shouldn't this `if (clientIterator->DataReceiverThreadId > 0)` be `if (clientIterator->DataReceiverThreadId >= 0)` instead, since the returned thread id from `vtkMultiThreader::SpawnThread` can be 0?

**Comment by lassoan** [comment]
Yes, good point, it should be >=0. The previous implementation just checked for >0 but that was incorrect (everywhere else the check is for >=0).

**Comment by adamaji** [comment]
@lassoan I think this issue can now be closed with #720 in?

**Comment by lassoan** [comment]
Yes, thank you!
